### PR TITLE
Integrate mypy for Python type checking

### PR DESCRIPTION
**Pull request by quitterie-lcs**
## Purpose

It has been decided that we use type checking in ralph project.
[`mypy`](https://mypy.readthedocs.io/en/stable/index.html) is the best candidate for ralph.

## Proposal

- [x] Add local tooling for development
- [x] Add CI tooling
- [x] Fix `mypy` warning

## Notes

All linter warnings are not solved in this issue.

### Models
- Unsolve valid-type for constrained type from Pydantic V1
- Unsolve assignment types for Pydantic model inheritences

### Backends
- Methods signatures in backends need redefinition considering types unions that are not compatible with mypy (particularly for query methods).

### API
 - Fix remaining issues for router statements and authentication

## Tips

Help for recurring problems:

- https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-new-additions-to-the-typing-module
- https://mypy.readthedocs.io/en/stable/kinds_of_types.html#tuple-types
- https://github.com/pydantic/pydantic/blob/50bc0039185c1522174c471bf967fd29a61fce74/pydantic/v1/dataclasses.py#L77
- https://github.com/hauntsaninja/no_implicit_optional
- Variable redefinition: https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-each-name-is-defined-once-no-redef 
- Missing imports (rfc3987 for e.g.): https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -125,17 +126,22 @@ exclude =
...
 skip_glob=venv
 profile=black
 
+[pydocstyle]
+convention = google
+match_dir = ^(?!tests|venv|build|scripts).*
+match = ^(?!(setup)\.(py)$).*\.(py)$
+
+[mypy]
+warn_return_any = True
```
Comment (wilbrdt) [inline setup.cfg:141]: Would a [disallow_untyped_defs = True](https://mypy.readthedocs.io/en/stable/config_file.html#confval-disallow_untyped_defs) be too much?

File: setup.cfg [hunk 2]
```diff
@@ -92,6 +93,10 @@ dev =
     pytest-httpx<0.23.0 # pin as Python 3.7 and 3.8 is no longer supported from release 0.23.0
     requests-mock==1.11.0
     responses<0.23.2 # pin until boto3 supports urllib3>=2
+    types-python-dateutil == 2.8.19.14 # Typing stubs for python-dateutil
+    types-python-jose == 3.3.4.8 # Typing stubs for python-jose
+    types-requests<2.31.0.7 # Typing stubs for requests
+    types-cachetools == 5.3.0.6 # Typing stubs for cachetools
```
Comment (jmaupetit) [inline setup.cfg:99]: I would remove those comments.

File: setup.cfg [hunk 3]
```diff
@@ -134,17 +139,50 @@ exclude =
...
+[mypy-ovh.*]
+ignore_missing_imports = True
+
+[mypy-swiftclient.service.*]
+ignore_missing_imports = True
+
+[pydantic-mypy]
+init_forbid_extra = True
+init_typed = True
+#warn_required_dynamic_aliases = True
```
Comment (jmaupetit) [inline setup.cfg:184]: An hesitation?

File: src/ralph/api/__init__.py [hunk 4]
```diff
@@ -14,12 +15,12 @@
...
 
 @lru_cache(maxsize=None)
-def get_health_check_routes():
+def get_health_check_routes() -> List:
     """Return the health check routes."""
     return [route.path for route in health.router.routes]
 
 
-def filter_transactions(event, hint):  # pylint: disable=unused-argument
+def filter_transactions(event: Any, hint) -> Any:  # pylint: disable=unused-argument
```
Comment (jmaupetit) [inline src/ralph/api/__init__.py:23]: Maybe we can be more specific?

```suggestion
def filter_transactions(event: Dict, hint) -> Dict:  # pylint: disable=unused-argument
```

**Review by wilbrdt — APPROVED** [review]
CI tooling would be added in a future PR, right?
Otherwise, looks good to me as a first step, well done! :tada:

File: src/ralph/parsers.py [hunk 5]
```diff
@@ -33,8 +34,8 @@ class GELFParser(BaseParser):
 
     name = "gelf"
 
-    def parse(self, input_file):
-        """Parse GELF formatted logs (one JSON string event per row).
+    def parse(self, input_file: Union[TextIO, BinaryIO]) -> Generator:
+        """Parses GELF formatted logs (one JSON string event per row).
```
Comment (Leobouloc) [inline src/ralph/parsers.py:38]:
```suggestion
        """Parse GELF formatted logs (one JSON string event per row).
```

File: setup.cfg (hunk 3 above)
Comment (quitterie-lcs) [reply setup.cfg:184]: No it is an oversight

File: src/ralph/api/auth/oidc.py [hunk 6]
```diff
@@ -53,7 +53,7 @@ class Config:  # pylint: disable=missing-class-docstring # noqa: D106
 
 
 @lru_cache()
-def discover_provider(base_url: AnyUrl) -> dict:
+def discover_provider(base_url: AnyUrl) -> Any:
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/api/auth/oidc.py:56]: Returning `response.json()` is considered as returning `Any` for `mypy`

File: src/ralph/api/auth/oidc.py [hunk 7]
```diff
@@ -71,7 +71,7 @@ def discover_provider(base_url: AnyUrl) -> dict:
 
 
 @lru_cache()
-def get_public_keys(jwks_uri: AnyUrl) -> dict:
+def get_public_keys(jwks_uri: AnyUrl) -> Any:
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/api/auth/oidc.py:74]: Please refer to the answer of the upper comment

File: src/ralph/models/xapi/base/unnested_objects.py [hunk 8]
```diff
@@ -81,7 +82,7 @@ class BaseXapiActivityInteractionDefinition(BaseXapiActivityDefinition):
 
     @validator("choices", "scale", "source", "target", "steps")
     @classmethod
-    def check_unique_ids(cls, value):
+    def check_unique_ids(cls, value: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/xapi/base/unnested_objects.py:85]: Well spotted, thank you!

File: src/ralph/models/converter.py [hunk 9]
```diff
@@ -34,7 +45,13 @@ class ConversionItem:
     transformers: Tuple[Callable[[Any], Any]]
     raw_input: bool
 
-    def __init__(self, dest: str, src=None, transformers=lambda _: _, raw_input=False):
+    def __init__(
+        self,
+        dest: str,
+        src: Union[str, None] = None,
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/converter.py:51]: Sounds better, thanks!

File: Makefile [hunk 10]
```diff
@@ -222,6 +222,11 @@ lint-pydocstyle: ## lint Python docstrings with pydocstyle
 	@$(COMPOSE_TEST_RUN_APP) pydocstyle
 .PHONY: lint-pydocstyle
 
+lint-mypy: ## lint back-end python sources with mypy
+	@echo 'lint:mypy started…'
+	@$(COMPOSE_TEST_RUN_APP) mypy
+.PHONY: lint-mypy
```
Comment (Leobouloc) [inline Makefile:228]: Maybe we should include this in `make lint` ?

File: src/ralph/utils.py [hunk 11]
```diff
@@ -153,7 +156,9 @@ def set_dict_value_from_path(dict_: dict, path: List[str], value: any):
     dict_[path[-1]] = value
 
 
-async def gather_with_limited_concurrency(num_tasks: Union[None, int], *tasks):
+async def gather_with_limited_concurrency(
+    num_tasks: Union[None, int], *tasks: Any
```
Comment (Leobouloc) [inline src/ralph/utils.py:160]:
```suggestion
    num_tasks: Optional[int], *tasks: Any
```

**Review by Leobouloc — APPROVED** [review]
Looks good to me !!

File: Makefile (hunk 10 above)
Comment (quitterie-lcs) [reply Makefile:228]: As all warnings are not yet fixed, I voluntarily did not include in the linter to be consistent with CI linter

**Pull request merged**